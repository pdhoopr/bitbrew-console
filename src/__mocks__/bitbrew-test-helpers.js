import diff from 'jest-diff';
import ShallowRenderer from 'react-test-renderer/shallow';

export function diffComponents(baseComponent, compareComponent) {
  const changes = diff(baseComponent, compareComponent, {
    aAnnotation: `<BaseComponent />`,
    bAnnotation: `<CompareComponent />`,
    contextLines: 3,
    expand: false,
  });
  return `Diff:\n${changes}`;
}

export function shallow(component) {
  const renderer = new ShallowRenderer();
  renderer.render(component);
  return renderer.getRenderOutput();
}

export const fixtures = {
  accessToken:
    'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJPbmxpbmUgSldUIEJ1aWxkZXIiLCJpYXQiOjE1MjUxMTUzNTAsImV4cCI6MTU1NjY1MTM1MCwiYXVkIjoid3d3LmV4YW1wbGUuY29tIiwic3ViIjoianJvY2tldEBleGFtcGxlLmNvbSJ9.4O-W8aeOOUFU8fkGTAFaAu9D9nK_l61yPomNG1Qkny8',
  orgs: {
    danlaw: {
      id: 'bd006616-3bd2-4783-a84d-b0fa4ed81135',
      properName: 'Danlaw',
    },
    azuga: {
      id: '8d9ec637-f3bb-4c46-83ee-11764d09f0b2',
      properName: 'Azuga',
    },
    bitbrew: {
      id: '94d9b0ca-a758-4502-8dcd-41d883e38714',
      properName: 'BitBrew',
    },
  },
  users: {
    bill: {
      email: 'bill@bitbrew.com',
      name: 'Bill Monaghan',
    },
    cynthia: {
      email: 'cynthia@bitbrew.com',
      name: 'Cynthia Hamilton',
    },
    alex: {
      email: 'bill@bitbrew.com',
      name: 'Alex Li',
    },
  },
};
