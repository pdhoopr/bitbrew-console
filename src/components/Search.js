import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';
import { createIdForA11y } from '../utils/tools';
import { Input } from './Forms';
import { SearchIcon } from './Icons';

const Wrapper = styled.div`
  display: block;
  flex: 1;
  position: relative;
`;

const Icon = styled(SearchIcon)`
  left: var(--size-14);
  pointer-events: none;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  z-index: 1;
`;

const Field = styled(Input)`
  background-color: var(--color-white);
  border: none;
  border-radius: var(--corner-radius);
  box-shadow: var(--elevation-low);
  padding-bottom: var(--size-8);
  padding-left: var(--size-48);
  padding-top: var(--size-8);
  transition: box-shadow var(--duration-short);

  &:hover,
  &:focus {
    box-shadow: var(--elevation-low-darker);
  }
`;

class Search extends React.Component {
  descriptionId = createIdForA11y(`${Search.name}__description`);

  render() {
    const { description, onChange, placeholder, value } = this.props;
    return (
      <Wrapper>
        <label htmlFor="query">
          <Icon aria-hidden />
          <Field
            id="query"
            value={value}
            onChange={onChange}
            placeholder={placeholder}
            aria-label={placeholder}
            aria-describedby={this.descriptionId}
          />
        </label>
        <p id={this.descriptionId} hidden>
          {description}
        </p>
      </Wrapper>
    );
  }
}

Search.propTypes = {
  description: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.string.isRequired,
};

Search.defaultProps = {
  placeholder: '',
};

export default observer(Search);
