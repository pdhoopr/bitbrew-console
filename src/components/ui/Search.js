import PropTypes from "prop-types";
import React, { useRef } from "react";
import styled from "styled-components";
import { generateId } from "../../utils";
import { Input } from "./Forms";
import { SearchIcon } from "./Icons";

const Wrapper = styled.div`
  display: block;
  flex: 1;
  position: relative;
`;

const Icon = styled(SearchIcon)`
  left: var(--size-14);
  pointer-events: none;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  z-index: 1;
`;

const Field = styled(Input)`
  background-color: var(--color-white);
  border: none;
  border-radius: var(--corner-radius);
  box-shadow: var(--elevation-low);
  padding-bottom: var(--size-8);
  padding-left: var(--size-48);
  padding-top: var(--size-8);
  transition: box-shadow var(--duration-short);

  &:hover,
  &:focus {
    box-shadow: var(--elevation-low-darker);
  }
`;

export default function Search({ description, onChange, placeholder, value }) {
  const descriptionId = useRef(generateId(`${Search.name}__description`));

  return (
    <Wrapper>
      <label htmlFor="searchTerm">
        <Icon aria-hidden />
        <Field
          id="searchTerm"
          value={value}
          onChange={onChange}
          placeholder={placeholder}
          aria-label={placeholder}
          aria-describedby={descriptionId.current}
        />
      </label>
      <p id={descriptionId.current} hidden>
        {description}
      </p>
    </Wrapper>
  );
}

Search.propTypes = {
  description: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  value: PropTypes.string.isRequired,
};

Search.defaultProps = {
  placeholder: "",
};
