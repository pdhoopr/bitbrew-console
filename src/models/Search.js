import { getEnv, types } from 'mobx-state-tree';

export default types
  .model('Search', {
    query: types.optional(types.string, ''),
  })
  .views(self => ({
    get results() {
      return getEnv(self).onSearch(self.query);
    },
  }))
  .actions(self => ({
    update(event) {
      self.query = event.currentTarget.value;
    },
  }));
