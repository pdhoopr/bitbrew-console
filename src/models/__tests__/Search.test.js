import { fixtures } from 'bitbrew-test-helpers';
import { getSnapshot } from 'mobx-state-tree';
import Search from '../Search';

test('initializes search with the default values', () => {
  const model = Search.create();
  expect(getSnapshot(model)).toMatchSnapshot();
});

test('initializes search with the defined values', () => {
  const data = {
    query: fixtures.orgs.bitbrew.properName,
  };
  const model = Search.create(data);
  expect(model).toMatchObject(data);
});

test('updates the value of the search query', () => {
  const query = fixtures.orgs.danlaw.properName;
  const model = Search.create();
  expect(model.query).not.toBe(query);
  model.update({
    currentTarget: {
      id: 'query',
      value: query,
    },
  });
  expect(model.query).toBe(query);
});

test('gets the search results using the query and a callback defined in the environment', () => {
  const env = {
    onSearch: (query) =>
      Object.values(fixtures.orgs)
        .filter((org) =>
          org.properName.toLowerCase().includes(query.toLowerCase()),
        )
        .sort((a, b) => a.properName.localeCompare(b.properName)),
  };
  const model = Search.create({}, env);
  model.update({
    currentTarget: {
      id: 'query',
      value: 'a',
    },
  });
  expect(model.results).toEqual([fixtures.orgs.azuga, fixtures.orgs.danlaw]);
});
