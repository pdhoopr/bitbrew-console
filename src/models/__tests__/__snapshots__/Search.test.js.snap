// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializes search with the default values 1`] = `
Object {
  "query": "",
}
`;
